mountinfo.go

```diff
@@ -89,57 +75,76 @@ func parseMountInfo(r io.Reader) ([]*MountInfo, error) {
 	}
 
 	mount := &MountInfo{
-		MajorMinorVer:  getStringSliceElement(beforeFields, 2, ""),
-		Root:           getStringSliceElement(beforeFields, 3, ""),
-		MountPoint:     getStringSliceElement(beforeFields, 4, ""),
-		Options:        mountOptionsParser(getStringSliceElement(beforeFields, 5, "")),
+		MajorMinorVer:  mountInfo[2],
+		Root:           mountInfo[3],
+		MountPoint:     mountInfo[4],
+		Options:        mountOptionsParser(mountInfo[5]),
 		OptionalFields: nil,
-		FSType:         getStringSliceElement(afterFields, 0, ""),
-		Source:         getStringSliceElement(afterFields, 1, ""),
-		SuperOptions:   mountOptionsParser(getStringSliceElement(afterFields, 2, "")),
+		FSType:         mountInfo[mountInfoLength-3],
+		Source:         mountInfo[mountInfoLength-2],
+		SuperOptions:   mountOptionsParser(mountInfo[mountInfoLength-1]),
 	}
 
-	mount.MountId, err = strconv.Atoi(getStringSliceElement(beforeFields, 0, ""))
+	mount.MountId, err = strconv.Atoi(mountInfo[0])
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse mount ID")
 	}
-	mount.ParentId, err = strconv.Atoi(getStringSliceElement(beforeFields, 1, ""))
+	mount.ParentId, err = strconv.Atoi(mountInfo[1])
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse parent ID")
 	}
 	// Has optional fields, which is a space separated list of values.
 	// Example: shared:2 master:7
-	if len(beforeFields) > 6 {
-		mount.OptionalFields = make(map[string]string)
-		optionalFields := beforeFields[6:]
-		for _, field := range optionalFields {
-			optionSplit := strings.Split(field, ":")
-			target, value := optionSplit[0], ""
-			if len(optionSplit) == 2 {
-				value = optionSplit[1]
-			}
-			// Checks if the 'keys' in the optional fields in the mountinfo line are acceptable.
-			// Allowed 'keys' are shared, master, propagate_from, unbindable.
-			if _, ok := validOptionalFields[target]; ok {
-				mount.OptionalFields[target] = value
-			}
+	if mountInfo[6] != "" {
+		mount.OptionalFields, err = mountOptionsParseOptionalFields(mountInfo[6 : mountInfoLength-4])
+		if err != nil {
+			return nil, err
 		}
 	}
 	return mount, nil
 }
 
+// mountOptionsIsValidField checks a string against a valid list of optional fields keys.
+func mountOptionsIsValidField(s string) bool {
+	switch s {
+	case
+		"shared",
+		"master",
+		"propagate_from",
+		"unbindable":
+		return true
+	}
+	return false
+}
+
+// mountOptionsParseOptionalFields parses a list of optional fields strings into a double map of strings.
+func mountOptionsParseOptionalFields(o []string) (map[string]string, error) {
+	optionalFields := make(map[string]string)
+	for _, field := range o {
+		optionSplit := strings.SplitN(field, ":", 2)
+		value := ""
+		if len(optionSplit) == 2 {
+			value = optionSplit[1]
+		}
+		if !mountOptionsIsValidField(optionSplit[0]) {
+			return optionalFields, fmt.Errorf("invalid optional key: %s", optionSplit[0])
```

Comment:
I'm wondering if we should just ignore invalid fields instead of returning an error since the procfs docs say: `Parsers should ignore all unrecognised optional fields.`

Reply:
@SuperQ Whats the reasoning for validating them at all? I feel like we should just include whatever is in there without any validation.

Reply:
Hmm, good point, I will update this part.

Reply:
@discordianfish I feel like we should follow the procfs docs and ignore them.